modules/packages/LinearAlgebra.chpl

```diff
@@ -1042,6 +1042,190 @@ proc trace(A: [?D] ?eltType) {
+  is a permutation matrix. 
+  
+  `L` and `U` are stored in the same matrix `LU` where 
+  the unit diagonal elements of L are not stored.
+  
+  `ipiv` contains the pivot indices such that row i of `A` 
+  was interchanged with row `ipiv(i)`.
+  
+*/
+
+proc lu (A: [?Adom] ?eltType) {
+  if Adom.rank != 2 then
+    halt("Wrong rank for LU factorization");
+
+  if Adom.shape(1) != Adom.shape(2) then
+    halt("LU factorization only supports square matrices");
+
+  var (LU, ipiv, numSwap) = _lu(A);
+  return (LU,ipiv);
+}
+
+proc permute (ipiv, A: [?Adom] ?eltType, transpose=false) {
+  const n = Adom.shape(1);
+  
+  var B: [Adom] eltType;
+  
+  if Adom.rank == 1 {
+    if transpose {
+      forall i in 1..n {
+        B[i] = A[ipiv(i)];
+      }
+    }
+    else {
+      forall i in 1..n {
+        B[ipiv(i)] = A[i];
+      }
+    }
+  }
+  else if Adom.rank == 2 {
+    if transpose {
+      forall i in 1..n {
+        B[i,..] = A[ipiv(i),..];
+      }
+    }
+    else {
+      forall i in 1..n {
+        B[ipiv(i),..] = A[i,..];
+      }
+    }
+  }
+  return B;
+}
+
+/* Returns the determinant of a square matrix.
+
+    .. note::
+
+      This procedure performs LU factorization to compute the 
+      determinant. In certain cases, e.g. having a lower/upper
+      triangular matrix, it is more desirable to compute the 
+      determinant manually.
+*/
+
+proc det (A: [?Adom] ?eltType) {
+  if Adom.rank != 2 then
+    halt("Wrong rank for computing determinant");
+
+  if Adom.shape(1) != Adom.shape(2) then
+    halt("Determinant can only be computed from square matrices");
+
+  var (LU,ipiv,numSwap) = _lu(A);
+  const pdet = if numSwap % 2 == 0 then 1 else -1;
+
+  // L[i,i] always = 1, so we only need to take the 
+  // diagonal product of U
+
+  return (* reduce [i in Adom.dim(1)] LU[i,i]) * pdet;
+}
+
+/* Returns the solution ``x`` to the linear system `` L * x = b `` 
+    where ``L`` is a lower triangular matrix. `unit_diag` will set the diagonal
+    elements to 1 within this procedure.
+*/
+proc solve_tril (const ref L: [?Ldom] ?eltType, const ref b: [?bdom] eltType, 
+                  unit_diag = true) {
+  const n = Ldom.shape(1);
+  var y = b;
+  
+  for i in 1..n {
+    const sol = if unit_diag then y(i) else y(i) / L(i,i);
+    y(i) = sol;
+    
+    if (i < n) {
+      forall j in (i+1)..n {
+        y(j) -= L(j,i) * sol;
+      }
+    }
+  }
+  
+  return y;
+}
+
+/* Returns the solution ``x`` to the linear system `` U * x = b `` 
+    where ``U`` is an upper triangular matrix.
+*/
+proc solve_triu (const ref U: [?Udom] ?eltType, const ref b: [?bdom] eltType) {
+  const n = Udom.shape(1);
+  var y = b;
+  
+  for i in 1..n by -1 {
+    const sol = y(i) / U(i,i);
+    y(i) = sol;
+    
+    if (i > 1) {
+      forall j in 1..(i-1) by -1 {
+        y(j) -= U(j,i) * sol;
+      }
+    }
+  }
+  
+  return y;
+}
+
+/* Returns the solution ``x`` to the linear system `` A * x = b ``.
```

Comment:
Apparently sphinx doesn't enjoy seeing `<space>``.` Removing spaces around `A*X=B` or removing the full period makes the doc compilable. It doesn't seem like the existing docs are very consistent, so I don't have a strong preference. But while there you should change the tense from present to imperative. i.e. not `Returns` but `Return`. Similarly, you should make that change for other functions, as well.